lib/rendering/shaders/composite.frag

```diff
@@ -56,22 +79,61 @@ vec4 sampleLayerYUYV(vec2 uv, int layer, vec4 dve, vec4 data) {
 	return vec4(col.r, col.g, col.b, a);
 }
 
-vec4 sampleLayerRGBA(vec2 uv, int layer, vec4 dve, vec4 data) {
-	vec4 col = texture(tex[layer*3], (uv / dve.z) - (dve.xy / dve.zw));
+vec4 sampleLayerRGBA(vec2 uv, uint src_idx, vec4 dve, vec4 data) {
+	vec4 col = texture(tex[src_idx*3], (uv / dve.z) - (dve.xy / dve.zw));
 	col.a *= data.x;
 	return col;
 }
 
-vec4 sampleLayerRGB(vec2 uv, int layer, vec4 dve, vec4 data) {
-	vec4 col = texture(tex[layer*3], (uv / dve.z) - (dve.xy / dve.zw));
+vec4 sampleLayerRGB(vec2 uv, uint src_idx, vec4 dve, vec4 data) {
+	vec4 col = texture(tex[src_idx*3], (uv / dve.z) - (dve.xy / dve.zw));
 	col.a = 1.0;
 	return col;
 }
 
+vec4 sampleLayerFallback(vec2 uv, vec4 dve) {
+	vec2 tpos = (uv / dve.z) - (dve.xy / dve.zw);
+
+	vec4 col = fallbackColour();
+
+	if(tpos.x < 0 || tpos.x > 1.0) {
+		col.a = 0.0;
+	}
+	if(tpos.y < 0 || tpos.y > 1.0) {
+		col.a = 0.0;
+	}
+
+	return col;
+}
+
+vec4 sampleLayer(vec2 uv, int src_idx, vec4 dve, vec4 data, uint srcType) {
+	// return sampleLayerDebugBBox(uv, src_idx, dve, data);
+	if (src_idx >= 0 && srcType == 0) { // YUV422Frames
```

Comment:
Not all gpu drivers will optimize this automatically so it's probably better to have an `if(src_idx >= 0 )` wrapping all the cases instead of testing it multiple times